src/acl.c

```diff
@@ -1965,6 +2405,120 @@ void addACLLogEntry(client *c, int reason, int context, int argpos, sds username
  * ACL related commands
  * ==========================================================================*/
 
+/* Add the formatted response from a single selector to the ACL GETUSER
+ * response. This function returns the number of fields added. 
+ * 
+ * Setting verbose to 1 means that the full qualifier for key and channel
+ * permissions are shown.
+ */
+int aclAddReplySelectorDescription(client *c, aclSelector *s) {
+    /* Commands */
+    addReplyBulkCString(c,"commands");
+    sds cmddescr = ACLDescribeSelectorCommandRules(s);
+    addReplyBulkSds(c,cmddescr);
+
+    rax *keyperms = raxNew();
+    
+    /* Key patterns */
+    addReplyBulkCString(c,"keys");
+    if (s->flags & SELECTOR_FLAG_ALLKEYS) {
+        addReplyArrayLen(c,1);
+        addReplyBulkCBuffer(c,"*",1);
+        raxInsert(keyperms, (unsigned char*) "~*", 2, NULL, NULL);
+    } else {
+        addReplyArrayLen(c,listLength(s->patterns));
+        listIter li;
+        listNode *ln;
+        listRewind(s->patterns,&li);
+        while((ln = listNext(&li))) {
+            keyPattern *pattern = (keyPattern *) listNodeValue(ln);
+            addReplyBulkCBuffer(c,pattern->pattern, sdslen(pattern->pattern));
+            /* Add the key pattern here sorted by the permissions */
+            sds pstr = sdsCatPatternString(sdsempty(), pattern);
+            raxInsert(keyperms, (unsigned char*) pstr, sdslen(pstr), NULL, NULL);
+            sdsfree(pstr);
+        }
+    }
+
+    /* Key patterns */
+    addReplyBulkCString(c,"key-permissions");
```

Comment:
Codewise this feels over-engineered, but it will make it easy to add more types of permissions in the future if we go down that route.

Reply:
@itamarhaber i'd like your feedback here too.
see discussion in https://github.com/redis/redis/pull/9974#issuecomment-1015318619 and https://github.com/redis/redis/pull/9974#issuecomment-1015660610

i see in the ACL GETUSER [docs](https://redis.io/commands/acl-getuser) that it says the `Note that command rules are returned as a string in the same format used with the ACL SETUSER command`, while the `keys` field is a plain list of key patterns (without the `~` DSL prefix),
so we can't use it now, and we need to choose between keeping this field (a list of all key that can be possibly accessed), or deleting / deprecating that field.

and in addition to that, we need to decide between adding a new field which has the `key-permissions` as DSL, or as currently implemented, a nested list of different permissions, each containing a list of patterns.

WDYT?

Reply:
Given that it's a new major release and this command is mostly user-facing, I say we can probably break compatibility and for every key return a tuple of (permissions, key pattern). Clients should be able to deal with both cases easily (depending on the type of every key element returned).

Reply:
@madolson i have some concern about the currently implemented approach.
the one we have now is that we generate 3 nested entries in the `key-permissions` map elements.
one for `read`, one for `write`, one for `all`.
but what if we add more combinations in the future (insert, delete, etc)
instead of 3 options (mask of 2 bits), we may get 7 combinations.
i'm leaning towards making the `key-permissions` field a DSL like `commands`
WDYT?

Reply:
@oranagra Do mean a single string like commands (e.g. `~foo ~bar` as a single argument) or still split into an array of individual DSL arguments? I prefer the second, which would look like:
```
 7) "keys"
 8) 1) "~foo"
    2) "~bar"
```

Reply:
Let's try to imitate the `commands` field. 
I must say I don't understand why he chose to do the keys this way.. 
Also, considering that GETUSER is the opposite of SETUSER, maybe it should have been one long dsl string, and not resp. 

But take my advice with a grain of salt, I didn't play much with the ACL commands.

Reply:
Sounds good enough. Let's just make it all look the same.